Tests/OptimizelyTests-DataModel/OptimizelyJSONTests.swift

```diff
@@ -0,0 +1,489 @@
+        
+        // Verifying json2 toString was valid by converting it to dict and comparing with json1 dict
+        let json2StringToMap = OTUtils.convertToDictionary(text: optimizelyJSON2ToString!)
+        XCTAssertNotNil(json2StringToMap)
+        XCTAssertTrue(NSDictionary(dictionary: optimizelyJSON1ToMap).isEqual(to: json2StringToMap!))
+    }
+    
+    func testConstructorWithInvalidPayloadString() {
+        var err: Error?
+        var optimizelyJSON: OptimizelyJSON?
+        do {
+            try optimizelyJSON = OptimizelyJSON(payload: "incorrect_string")
+        } catch {
+            err = error
+        }
+        XCTAssertEqual(err?.localizedDescription, OptimizelyError.failedToConvertStringToDictionary.reason)
+        XCTAssertNil(optimizelyJSON)
+    }
+    
+    func testConstructorWithInvalidJSONArrayPayloadString() {
+        let jsonArrayPayload = """
+        [{
+        "field1": 1,
+        "field2": 2.5,
+        "field3": "three",
+        "field4": {"inner_field1":3,"inner_field2":["1","2",3.01,4.23,true]},
+        "field5": true,
+        }]
+        """
+        var err: Error?
+        var optimizelyJSON: OptimizelyJSON?
+        do {
+            try optimizelyJSON = OptimizelyJSON(payload: jsonArrayPayload)
+        } catch {
+            err = error
+        }
+        XCTAssertEqual(err?.localizedDescription, OptimizelyError.failedToConvertStringToDictionary.reason)
+        XCTAssertNil(optimizelyJSON)
+    }
+    
+    func testConstructorWithNonSerializableData() {
+        let data: [String: Any] = ["test": UIImage()]
+        var err: Error?
+        do {
+            try _ = OptimizelyJSON(data: data)
+        } catch {
+            err = error
+        }
+        XCTAssertEqual(err?.localizedDescription, OptimizelyError.invalidDictionary.reason)
+    }
+    
+    func testToMap() {
+        var err: Error?
+        var optimizelyJSONToMap: [String:Any]!
+        do {
+            try optimizelyJSONToMap = self.optimizelyJSON.toMap()
+        } catch {
+            err = error
+        }
+        XCTAssertNil(err)
+        XCTAssertTrue(NSDictionary(dictionary: optimizelyJSONToMap).isEqual(to: self.data))
+    }
+    
+    func testToString() {
+        var err: Error?
+        var optimizelyJSONToString: String!
+        do {
+            try optimizelyJSONToString = self.optimizelyJSON.toString()
+        } catch {
+            err = error
+        }
+        XCTAssertNil(err)
+        XCTAssertEqual(optimizelyJSONToString, self.payload)
+    }
+    
+    func testGetValueForInvalidJSONKeyAndEmptySchema() {
+        var err: Error?
+        var schema = EmptySchema()
+        do {
+            try self.optimizelyJSON.getValue(jsonPath: "field4.", schema: &schema)
+        } catch {
+            err = error
+        }
+        XCTAssertEqual(err?.localizedDescription, OptimizelyError.valueForKeyNotFound("").reason)
```

Comment:
is it the appropriate message? Question for Pawel/jae.

Reply:
Since we explode jsonPath to array using `.` separator, `field4.` would result into an empty string on first index of array which will produce the above error with empty `key` value.